src/Runtime/Core/Collections/ImmutableArray.cs

```diff
@@ -61,32 +71,68 @@ public ReadOnlyArray(T[] array)
         /// <remarks>
         /// This is value-type non-virtual version of enumerator that is used by compiler in foreach loops.
         /// </remarks>
-        public ReadOnlyArraySegmentEnumerator<T> GetEnumerator() =>
-            new ReadOnlyArraySegmentEnumerator<T>(this, 0, this.array.Length);
+        public ImmutableArraySegmentEnumerator<T> GetEnumerator() =>
+            new ImmutableArraySegmentEnumerator<T>(this, 0, this.array.Length);
 
         /// <inheritdoc/>
         IEnumerator<T> IEnumerable<T>.GetEnumerator() =>
-            new ReadOnlyArraySegmentEnumerator<T>(this, 0, this.array.Length);
+            new ImmutableArraySegmentEnumerator<T>(this, 0, this.array.Length);
 
         /// <inheritdoc/>
         IEnumerator IEnumerable.GetEnumerator() =>
-            new ReadOnlyArraySegmentEnumerator<T>(this, 0, this.array.Length);
+            new ImmutableArraySegmentEnumerator<T>(this, 0, this.array.Length);
 
-        /// <summary>
-        /// Helper method which allows to cast regular arrays to read only versions implicitly.
-        /// </summary>
-        public static implicit operator ReadOnlyArray<T>(T[] array) => new ReadOnlyArray<T>(array);
+        public override bool Equals(object o) => o is ImmutableArray<T> that && this == that;
+
+        public override int GetHashCode() => this.array.GetHashCode();
+
+        public static bool operator ==(ImmutableArray<T> left, ImmutableArray<T> right) =>
+            left.array == right.array;
+
+        public static bool operator !=(ImmutableArray<T> left, ImmutableArray<T> right) =>
+            left.array != right.array;
+
+        public struct Builder
+        {
+            private bool dirty;
+            private T[] array;
+
+            public Builder(int size)
+            {
+                this.dirty = false;
+                this.array = new T[size];
+            }
+
+            public T this[int index]
+            {
+                get => this.array[index];
+                set => this.array[index] = value;
```

Comment:
Should throw if dirty=true because the array now belongs to someone else.

Reply:
Removed dirty flag entirely